runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowExecutionStateSampler.java

```diff
@@ -0,0 +1,136 @@
+  private static final DataflowExecutionStateSampler INSTANCE =
+      new DataflowExecutionStateSampler(SYSTEM_MILLIS_PROVIDER);
+
+  private final ConcurrentHashMap<String, DataflowExecutionStateTracker> activeTrackersByWorkId =
+      new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<String, Map<String, IntSummaryStatistics>>
+      completedProcessingMetrics = new ConcurrentHashMap<>();
+
+  public static DataflowExecutionStateSampler instance() {
+    return INSTANCE;
+  }
+
+  @VisibleForTesting
+  public static DataflowExecutionStateSampler newForTest(MillisProvider clock) {
+    return new DataflowExecutionStateSampler(checkNotNull(clock));
+  }
+
+  public DataflowExecutionStateSampler(MillisProvider clock) {
+    super(clock);
+  }
+
+  @Override
+  public void addTracker(ExecutionStateTracker tracker) {
+    if (!(tracker instanceof DataflowExecutionStateTracker)) {
+      return;
+    }
+    DataflowExecutionStateTracker dfTracker = (DataflowExecutionStateTracker) tracker;
+    this.activeTrackersByWorkId.put(dfTracker.getWorkItemId(), dfTracker);
+  }
+
+  private static Map<String, IntSummaryStatistics> mergeStepStatsMaps(
+      Map<String, IntSummaryStatistics> map1, Map<String, IntSummaryStatistics> map2) {
+    for (Entry<String, IntSummaryStatistics> steps : map2.entrySet()) {
+      map1.compute(
+          steps.getKey(),
+          (k, v) -> {
+            if (v == null) {
+              return steps.getValue();
+            }
+            v.combine(steps.getValue());
+            return v;
+          });
+    }
+    return map1;
+  }
+
+  @Override
+  public void removeTracker(ExecutionStateTracker tracker) {
+    if (!(tracker instanceof DataflowExecutionContext.DataflowExecutionStateTracker)) {
+      return;
+    }
+    DataflowExecutionStateTracker dfTracker = (DataflowExecutionStateTracker) tracker;
+    completedProcessingMetrics.put(dfTracker.getWorkItemId(), dfTracker.getProcessingTimesByStep());
+    activeTrackersByWorkId.remove(dfTracker.getWorkItemId());
+
+    // Attribute any remaining time since the last sampling while removing the tracker.
+    //
+    // There is a race condition here; if sampling happens in the time between when we remove the
+    // tracker from activeTrackers and read the lastSampleTicks value, the sampling time will
+    // be lost for the tracker being removed. This is acceptable as sampling is already an
+    // approximation of actual execution time.
+    long millisSinceLastSample = clock.getMillis() - this.lastSampleTimeMillis;
+    if (millisSinceLastSample > 0) {
+      tracker.takeSample(millisSinceLastSample);
+    }
+  }
+
+  @Override
+  public void doSampling(long millisSinceLastSample) {
+    for (DataflowExecutionStateTracker tracker : activeTrackersByWorkId.values()) {
+      tracker.takeSample(millisSinceLastSample);
+    }
+  }
+
+  @Nullable
+  public ActiveMessageMetadata getActiveMessageMetadataForWorkId(String workId) {
```

Comment:
return Optional<ActiveMessageMetadata>, we want to remove `null` from the API whereever it makes sense.

Reply:
[075dd57](https://github.com/apache/beam/pull/29592/commits/075dd57cca2409b576222ea6eb014f3dfe368bf0)